### PR TITLE
Feature/jf/save extra proj params

### LINKED ISSUE #213: Refactor chart removal
Chart removal is a little messy. Several chart-related settings have been added and are force- removed/reset alongside charts. If possible, refactor charts such that all its associated settings are contained on the chart so removal will reset all associated settings

Run into especially with #210

### PR DESCRIPTION
**Pull request by fungjj92**
## Overview

Lots of things!
There were some loose-ends from adding threshold params that caused strange behavior. This PR pulls in a few issues that tied together to complete the feature.

### Notes

I left comments in the code to explain some of the decisions I made to work with Angular's lifecycle hooks (https://angular.io/guide/lifecycle-hooks#lifecycle-hooks), which became a little tricky with parent-child communication and reactive forms. Happy to talk through anything via messages or in person and/or provide more robust code comments!

## Testing Instructions

Checkout CC API https://github.com/azavea/climate-change-api/pull/649 and run locally
Set your apiHost to localhost
`npm run serve`

Make sure all threshold indicators work when you:
- select them
- select another one
- select non-thresh indicators
- change thresh params
- save the project and come back (aka leave the view)
- change navbar settings

Closes #213, #209, #217

### CODE REVIEW

File: src/app/charts/extra-params-components/threshold.component.ts [hunk 1]
```diff
@@ -48,54 +50,62 @@ export class ThresholdComponent implements AfterViewInit, OnChanges {
...
-                'threshold': form.thresholdCtl,
-                'threshold_units': form.thresholdUnitCtl
-            }});
-        });
+    ngOnChanges(changes: any) {
+        /* Ignore change detection:
+            - before initalization
+            - from extraParams @Input whose initial values are all we want
+        */
+        if (changes.label && this.thresholdForm) {
```
Comment (CloudNiner) [inline src/app/charts/extra-params-components/threshold.component.ts]: Looked at this a bit more as promised. I think this is as good as it gets. Unfortunately, I'm still very concerned how extra logic like this in these template components will scale as we add additional ones, but that's a problem that's above the scale of this PR.

The only other alternative would be to switch the two explicit `@Input()` and `@Output` decorators that deal with `extraParams` to a single two way binding via `[()]`. But, this doesn't allow you to manually limit when we propagate changes up to the parent component. So it's no good. The `@Input` decorator doesn't have any additional options for ignoring changes after the first one, and in fact, that's not what you want here. You want the form to reset anytime the `label` property changes, which indirectly means the indicator this component is referring to, changes. 

Maybe all we do here is cleanup the comment to describe more the behavior we want, rather than just restating what the code is doing. e.g. switch the comment to something like "Only reset the form and push changes to the parent when the indicator changes via the label property, and when the form exists".

File: src/app/charts/extra-params-components/threshold.component.ts [hunk 2]
```diff
@@ -48,54 +50,62 @@ export class ThresholdComponent implements AfterViewInit, OnChanges {
 
     @Output() thresholdParamSelected = new EventEmitter<any>();
 
-    createForm() {
-        this.thresholdForm = this.fb.group({
-            comparatorCtl: [this.defaultComparator, Validators.required],
-            thresholdCtl: [this.defaultThreshold, Validators.required],
-            thresholdUnitCtl: [this.defaultTemperatureUnit, Validators.required]
-        });
+    constructor(private fb: FormBuilder) {}
```
Comment (CloudNiner) [inline src/app/charts/extra-params-components/threshold.component.ts]: As noted when looking at this with you, change to `private formBuilder: FormBuilder`

File: src/app/charts/extra-params-components/threshold.component.ts [hunk 3]
```diff
@@ -48,54 +50,62 @@ export class ThresholdComponent implements AfterViewInit, OnChanges {
...
+                comparatorCtl: this.defaultComparator,
+                thresholdCtl: this.defaultThreshold
+            });
+        }
     }
 
-    constructor(private fb: FormBuilder) {
-        this.createForm();
+    ngOnInit() {
+        this.createForm(); // must create form on init instead of constructor to capture @Input values
```
Comment (CloudNiner) [inline src/app/charts/extra-params-components/threshold.component.ts]: Comment on its own line

File: src/app/charts/extra-params-components/threshold.component.ts [hunk 4]
```diff
@@ -48,54 +50,62 @@ export class ThresholdComponent implements AfterViewInit, OnChanges {
...
+            this.thresholdParamSelected.emit({data: {
+                'event': event,
+                'threshold_comparator': form.comparatorCtl,
+                'threshold': form.thresholdCtl,
+                'threshold_units': form.thresholdUnitCtl
+            }});
+        });
+    }
+
+    private evaluateVariable() {
```
Comment (CloudNiner) [inline src/app/charts/extra-params-components/threshold.component.ts]: Since we need to know the default unit, maybe we should pass `Indicator` instead of `label` as an input, such that we could just pull the default unit from there rather than attempting to magically derive it based on substrings in the label.

File: src/app/charts/line-graph.component.ts [hunk 5]
```diff
@@ -79,7 +79,7 @@ export class LineGraphComponent implements OnChanges, AfterContentInit {
     }
 
     /* Executes on every @Input change */
-    ngOnChanges(): void {
+    ngOnChanges(changes?: any): void {
```
Comment (CloudNiner) [inline src/app/charts/line-graph.component.ts]: Was this change necessary for some reason? I don't think we need to redeclare the type of this method.

File: src/app/lab/lab.component.ts [hunk 6]
```diff
@@ -89,9 +93,13 @@ export class LabComponent implements OnInit, OnDestroy {
             return;
         }
         this.removeChart();
-        this.indicator = indicator;
-        const chart = new Chart({indicator: indicator,
-                                 unit: indicator.default_units});
-        this.project.project_data.charts = [chart];
+        // Trigger lifecycle evaluation to truly destroy the chart child component
```
Comment (CloudNiner) [inline src/app/lab/lab.component.ts]: This comment should probably have a bit more description of _why_ we want to completely destroy this component when we select a different indicator, so the context is apparent for the next person that looks at this.

File: src/app/charts/line-graph.component.ts (hunk 5 above)
Comment (fungjj92) [reply src/app/charts/line-graph.component.ts]: Leftover from playing around, good catch

File: src/app/charts/extra-params-components/threshold.component.ts (hunk 4 above)
Comment (fungjj92) [reply src/app/charts/extra-params-components/threshold.component.ts]: Hm upon review,`Indicator.default_units` doesn't necessarily coordinate to pr/temp unit (C,F,K,in,etc). It can be `days`for some indicators, so derivation is necessary in some form. I updated it to depend on `Indicator.variables,` which I think is more readable.

File: src/app/charts/extra-params-components/threshold.component.ts (hunk 1 above)
Comment (fungjj92) [reply src/app/charts/extra-params-components/threshold.component.ts]: I just realized that we never need to reset the form to defaults, since we destroy the threshold component along with the chart component between indicator selections. I removed `ngOnChanges()` and everything works the same, and we avoid constant listening for `@Input` -- whose constant listening depends on having `ngOnChanges()`

**Comment by fungjj92** [comment]
Updated

File: src/app/charts/extra-params-components/threshold.component.ts (hunk 4 above)
Comment (CloudNiner) [reply src/app/charts/extra-params-components/threshold.component.ts]: Nice catch.

File: src/app/lab/lab.component.ts [hunk 7]
```diff
@@ -89,9 +93,15 @@ export class LabComponent implements OnInit, OnDestroy {
             return;
         }
         this.removeChart();
-        this.indicator = indicator;
-        const chart = new Chart({indicator: indicator,
-                                 unit: indicator.default_units});
-        this.project.project_data.charts = [chart];
+        /*  Trigger lifecycle to truly destroy the chart component & its children
```
Comment (CloudNiner) [inline src/app/lab/lab.component.ts]: I still dislike the decision to destroy/recreate, as completely destroying components and rebuilding is heavy handed and leads to cascading troublesome design decisions, such as the latest comment about removing the ngOnChanges no longer being necessary because we destroy/create. If we were ever to refactor and remove this destroy/create cycle, then the child ThresholdComponent would be broken because it no longer watches for the `@Input indicator` changes. As we continue to build based on this implicit behavior, it becomes easier to make similar decisions in the future which would further compound this issue.

With that said, I understand the impetus for it and would like to see the remaining tasks unblocked. A way to simplify this may become clear as we work through those tasks.
